plover/oslayer/osxkeyboardcontrol.py

```diff
@@ -300,37 +249,89 @@ def characters(s):
+        key_plan = []
+
+        # apply_raw's properties are used to store the current keycode sequence,
+        # and add the sequence to the key plan when called as a function.
+        def apply_raw():
+            if hasattr(apply_raw, 'sequence') and len(apply_raw.sequence) is not 0:
+                apply_raw.sequence.extend(apply_raw.release_modifiers)
+                key_plan.append((self.RAW_PRESS, apply_raw.sequence))
+            apply_raw.sequence = []
+            apply_raw.release_modifiers = []
+        apply_raw()
+
+        last_modifier = None
         for c in characters(s):
-            event = CGEventCreateKeyboardEvent(MY_EVENT_SOURCE, 0, True)
-            set_string(event, c)
-            CGEventPost(kCGSessionEventTap, event)
-            event = CGEventCreateKeyboardEvent(MY_EVENT_SOURCE, 0, False)
-            set_string(event, c)
-            CGEventPost(kCGSessionEventTap, event)
+            for keycode, modifier in mac_keycode.KeyCodeForChar(c):
+                if keycode is not None:
+                    if modifier is not last_modifier:
+                        # Flush on modifier change.
+                        apply_raw()
+                        last_modifier = modifier
+                        modifier_keycodes = self._modifier_to_keycodes(modifier)
+                        apply_raw.sequence.extend(down(modifier_keycodes))
+                        apply_raw.release_modifiers.extend(up(modifier_keycodes))
+                    apply_raw.sequence.extend(down_up([keycode]))
+                else:
+                    # Flush on type change.
+                    apply_raw()
+                    key_plan.append((self.STRING_PRESS, c))
+        # Flush after string is complete.
+        apply_raw()
+
+        # We have a key plan for the whole string, grouping modifiers.
+        for press_type, sequence in key_plan:
+            if press_type is self.STRING_PRESS:
+                self._send_string_press(sequence)
+            elif press_type is self.RAW_PRESS:
+                self._send_sequence(sequence)
+
+    @staticmethod
+    def _send_string_press(c):
+        event = CGEventCreateKeyboardEvent(MY_EVENT_SOURCE, 0, True)
+        KeyboardEmulation._set_event_string(event, c)
+        CGEventPost(kCGSessionEventTap, event)
+        event = CGEventCreateKeyboardEvent(MY_EVENT_SOURCE, 0, False)
+        KeyboardEmulation._set_event_string(event, c)
+        CGEventPost(kCGSessionEventTap, event)
 
     def send_key_combination(self, combo_string):
         """Emulate a sequence of key combinations.
 
-        Argument:
-
-        combo_string -- A string representing a sequence of key
+        Args:
+            combo_string: A string representing a sequence of key
         combinations. Keys are represented by their names in the
         Xlib.XK module, without the 'XK_' prefix. For example, the
         left Alt key is represented by 'Alt_L'. Keys are either
```

Comment:
If think the convention we use elsewhere is right:

```python
    def send_key_combination(self, combo_string):
        """Emulate a sequence of key combinations.

        Arguments:

        combo_string -- A string representing a sequence of key combinations.
        Keys are represented by their names in the Xlib.XK module, without the
        'XK_' prefix. For example, the left Alt key is represented by 'Alt_L'.
        Keys are either separated by a space or a left or right parenthesis.
        Parentheses must be properly formed in pairs and may be nested. A key
        immediately followed by a parenthetical indicates that the key is
        pressed down while all keys enclosed in the parenthetical are pressed
        and released in turn. For example, Alt_L(Tab) means to hold the left
        Alt key down, press and release the Tab key, and then release the left
        Alt key.
        """
```

See [PEP-0257](https://www.python.org/dev/peps/pep-0257/#handling-docstring-indentation).